pkg/runner/test_runner.go

```diff
@@ -220,61 +225,51 @@ func (p *pipeline) route(terminateRoute <-chan struct{}, bundle test.TestStepBun
 		stepInClosed  bool
 		pendingTarget *target.Target
 		injectionWg   sync.WaitGroup
+		numIngress    uint64
 	)
 
 	for {
 		log.Debugf("listening")
+
 		select {
 		case <-terminateRoute:
 			err = fmt.Errorf("termination requested for routing into %s", stepLabel)
 			break
 		case injectionResult := <-injectResultCh:
 			log.Debugf("received injection result for %v", injectionResult.target)
-			ingressTarget[pendingTarget] = time.Now()
 			pendingTarget = nil
 			if injectionResult.err != nil {
 				err = fmt.Errorf("routing failed while injecting target %+v into %s", injectionResult.err, stepLabel)
 				targetInErrEv := testevent.Data{EventName: target.EventTargetInErr, Target: injectionResult.target}
 				if err := ev.Emit(targetInErrEv); err != nil {
 					log.Warningf("could not emit %v event for target: %+v", targetInErrEv, *injectionResult.target)
 				}
-				break
 			}
 			targetInEv := testevent.Data{EventName: target.EventTargetIn, Target: injectionResult.target}
 			if err := ev.Emit(targetInEv); err != nil {
 				log.Warningf("could not emit %v event for Target: %+v", targetInEv, *injectionResult.target)
 			}
-			if targets.Len() > 0 {
-				pendingTarget = targets.Back().Value.(*target.Target)
-				targets.Remove(targets.Back())
-				injectionWg.Add(1)
-				log.Debugf("writing target %v into test step", pendingTarget)
-				go targetWriter.writeTargetWithResult(terminate, pendingTarget, injectionChannels, &injectionWg)
-			}
+
 		case t, chanIsOpen := <-tRouteIn:
 			if !chanIsOpen {
-				// The previous routing block has closed our input channel, signaling that
-				// no more Targets will come through. Block reading from this channel
 				log.Debugf("routing input channel closed")
 				tRouteIn = nil
-			} else {
-				// Buffer the target and check if there is already an injection in progress.
-				// If so, pending targets will be dequeued only at the next result available
-				// on `injectResultCh`.
+				if position == 0 {
+					log.Debugf("setting in targets: %d", len(ingressTarget))
+					atomic.StoreUint64(&p.numIngress, numIngress)
```

Comment:
I am not sure I follow this: the idea here is that the first step of the pipeline should keep track of the number of targets injected, which at the moment it's done in a racy way via this `numIngress` (which is being replaced in a later PR by a channel). However the check for `position==0` will still be there, because the first injection is special and needs to do this additional accounting.

Reply:
Yeah, I understand the purpose of `position==0`, but I thought to apply another approach (discussed previously). But you explained that it won't work and moreover you said this logic is reworked in further PRs, so I just "resolve" this discussion thread :)